unixfs/tar/writer.go

```diff
@@ -32,16 +33,30 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
 	bufw := bufio.NewWriterSize(pipew, DefaultBufSize)
 
 	// construct the tar writer
-	w, err := NewWriter(bufw, dag, compression)
+	w, err := NewWriter(ctx, dag, archive, compression, bufw)
 	if err != nil {
 		return nil, err
 	}
 
 	// write all the nodes recursively
 	go func() {
-		if err := w.WriteNode(ctx, nd, filename); err != nil {
-			pipew.CloseWithError(err)
-			return
+		if !archive && compression != gzip.NoCompression {
+			// the case when the node is a file
+			dagr, err := uio.NewDagReader(w.ctx, nd, w.Dag)
+			if err != nil {
+				pipew.CloseWithError(err)
+				return
+			}
+
+			if _, err := dagr.WriteTo(w.w); err != nil {
```

Comment:
this is hacky-- the `not-tar` case does not really belong with the tar case. the "node is a file and no archive" case should be handled outside of this. maybe we can re-write this `DagArchive` function as:

```go
// DagArchive is equivalent to `ipfs getdag $hash | maybe_tar | maybe_gzip`
func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService, archive bool, compression int) (io.Reader, error) {

  _, filename := path.Split(name)

  // need to connect a writer to a reader
  piper, pipew := io.Pipe()

  // use a buffered writer to parallelize task
  bufw := bufio.NewWriterSize(pipew, DefaultBufSize)

  // write all the nodes recursively
  go func() {
    if !archive && compression != gzip.NoCompression {
      // the case when the node is a file.
      dagr, err := uio.NewDagReader(w.ctx, nd, w.Dag)
      if err != nil {
        pipew.CloseWithError(err)
        return
      }

      if _, err := dagr.WriteTo(bufw); err != nil {
        pipew.CloseWithError(err)
        return
      }

    } else {
      // construct the tar writer
      tarw, err := NewWriter(ctx, dag, archive, compression, bufw)
      if err != nil {
        return nil, err
      }

      if err := tarw.WriteNode(nd, filename); err != nil {
        pipew.CloseWithError(err)
        return
      }
    }

    if err := bufw.Flush(); err != nil {
      pipew.CloseWithError(err)
      return
    }

    w.Close()
    pipew.Close() // everything seems to be ok.
  }()

  return piper, nil
}
```

and pull it out of the `unixfs/tar` package into something else (maybe `unixfs/archive`?)